src/compiler/scala/tools/nsc/backend/jvm/analysis/NullnessAnalyzer.scala

```diff
@@ -191,11 +236,24 @@ class NullnessFrame(nLocals: Int, nStack: Int) extends AliasingFrame[NullnessVal
 
     super.execute(insn, interpreter)
 
-    if (nullCheckedAliases != null) {
-      val it = nullCheckedAliases.iterator
+    def setNullness(s: AliasSet, v: NullnessValue) = {
+      val it = s.iterator
       while (it.hasNext)
-        this.setValue(it.next(), NotNullValue)
+        this.setValue(it.next(), v)
+    }
+
+    if (ifNullAliases != null) {
+      val v = if (insn.getOpcode == IFNULL) NotNullValue else NullValue
+      postBranch = () => {
+        setNullness(ifNullAliases, v)
+        postBranch = () => {
+          setNullness(ifNullAliases, v.invert)
+          postBranch = null
+        }
+      }
```

Comment:
We could probably make this a static (non-capturing lambda) by, for instance, duplicating lines 247-253 within each side of the conditional on line 246.

Perhaps we could encode the states without lambdas to make things easier to debug? e.g. `data BranchNullState = None | Branch1NotNull | Branch1Null | Branch2Null | Branch2NotNull`.